ops/ruby/devops-toolkit/README.md

```diff
@@ -70,6 +70,14 @@ bundle install
 ## Tasks
 A variety of tasks (sub-commands) are available via the required `-t` option. Here  are their
 brief descriptions:
+- `backup-config`: Generate JSON or YAML files for each Stackdriver asset (other GCP assets TBD),
```

Comment:
This is separably from the implementation details of the tool, but where do you intend to archive these? It sounds like it's too large to want to put this into any of our git repos.

Reply:
So there was a discussion a while back where @gjuggler mentioned that have achieved "configuration as code" for Stackdriver, and I've been looking for a way to make that an actual reality. https://precisionmedicineinitiative.atlassian.net/browse/RW-4946 is the task for the periodic archival. The simplest approach is to just make a zip file and post it every day and delete the Nth oldest, so they just rotate.

If I can output straight to BigQuery that would be even better.

Reply:
My understanding of config-as-code is that the git repo is the canonical archive of config, same as our code. Maybe I'm missing something here, though. Is there something different about our Stackdriver monitoring setup that makes it necessary to pull things back down from Google and archive them?

Reply:
You have to have a systematic way of keeping up with it. The first step is mirroring what Stackdriver has locally, then archiving that, then restoring it. To put it into git is doable, though you'd want to strip out non-custom things more aggressively.

I'm not sure what form the "config" was expected to take if not this JSON or XML version. Either we start with that or build it with a template (which I may do next).

If the goal is to make the stored config the source of truth, then there needs to be some kind of reloading or two-way synchronizing happening periodically.

The work on the fetching pattern should be useful on the uploading step too though.

Reply:
If we want to put it in git, I can make a place in workbench-devops and do another PR there. I'd change the name structure to remove the multiple-files-per-config-named-after-timestamp and let git do the hard work on diffing, etc. (Sorting the arbitrarily ordered keys when writing the files will help here).

Having it directly in git is clearly the ideal from a config-as-code standpoint.

Moving from here to Terraform templates would be another increment of work.

Reply:
The "code-as-config" model would typically treat the code as the source of truth, with a one way sync from code -> deployed system. There would be no two-way sync. This simplification also allows you to templatize/parameterize these configs so that you're not managing N fully attenuated configs for N environments. Some one-time bootstrapping of these configs is fine, but that's probably a oneoff manual process.

I have the same question as Greg, which I don't think was answered above, which is why it is necessary to sync from Stackdriver -> code config, or maintain a mechanism for performing any kind of two-way sync.

Reply:
People like to fiddle with graphs and dashboards, and adjust levels and queries interactively. It would be very painful to lock down the console and only allow text edits.

With this tool, we can periodically (or right after a fiddling session) pull the configs, write them on top of what's in github, and PR the differences.

It would never have occurred to me to do a one-way only sync with an interactive, graphical tool (with such a relatively low-level API). Yes, the config is the source of truth, but prior to its existence we need a systematic way to pull it and arrange it.

Of course going up is just in the inverse of coming down. You still need to organize the config in some manner, and having it in the same form that the API wants is a plus. But since the input of the up direction is a subset of the output from the down direction, it was natural to do the down direction first. Plus, it let me get a handle on how much stuff we were working with.

Finally, this tool should work for every GCP project that uses Stackdriver, not just ours, so a self-bootstrapping capability is a big plus.

Reply:
Sorry, has the overall strategy you're describing already been discussed and agreed upon? Or are we having this discussion now from first principles? If this PR is aligned with what you and @gjuggler have already discussed, I won't block this. @gjuggler you can approve now if this is the case.

I don't think the code is the source of truth in the model you're describing. It's derived/pulled from the live config defined in the UI. If we're treating the UI as source of truth, I'd prefer not to check these configs into git as they must lag behind the UI by definition and are likely to go out of sync. If we want to drive all of the monitoring configs using a config-as-code approach, I think the way I'd expect this to work would be:
- prefer to isolate any "fiddling" to explicit personal/sandbox/experimental dashboards or to freeform query editors
- Use a query editor to build dashboards / alert policies
- When you're done, pull the JSON/yaml representation of this out, and convert it into a templated form as a PR, such that they are appropriately parameterized for all relevant environments
- Right before merging the PR, push the canonical configs to all environments
- You should expect any changes you make to the production configs outside of git (via the UI) may be blown away on the next monitoring config push

If we're using the UI as the source of truth, and periodically archiving our configs outside of source control, that sounds fine if that's the pragmatic way forward.